spanner/value.go

```diff
@@ -49,13 +50,55 @@ const (
 	NumericScaleDigits = 9
 )
 
+// NumericLossHandlingMode describes the mode of how to deal with loss of
+// precision on numeric values.
+type NumericLossHandlingMode int
+
+const (
+	// NumericRound rounds an invalid numeric value, e.g., 0.1234567895 rounds
+	// to 0.123456790.
+	NumericRound NumericLossHandlingMode = iota
+	// NumericError throws an error when meeting invalid numeric value.
+	NumericError
+)
+
+// NumericPrecisionLossHandling is the configuration for hanlding loss of
+// precission on numeric values.
+var NumericPrecisionLossHandling NumericLossHandlingMode
+
 // NumericString returns a string representing a *big.Rat in a format compatible
 // with Spanner SQL. It returns a floating-point literal with 9 digits after the
 // decimal point.
 func NumericString(r *big.Rat) string {
 	return r.FloatString(NumericScaleDigits)
 }
 
+// validateNumeric returns nil if there are no errors. It will return an error
+// when the numeric number is not valid.
+func validateNumeric(r *big.Rat) error {
+	if r == nil {
+		return nil
+	}
+	// Add one more digit to the scale component to find out if there are more
+	// digits than required.
+	strRep := r.FloatString(NumericScaleDigits + 1)
+	strRep = strings.TrimRight(strRep, "0")
+	strRep = strings.TrimLeft(strRep, "-")
+	s := strings.Split(strRep, ".")
+	if len(s) != 2 {
```

Comment:
Will this ever be true for any `big.Rat` value? If so, what happens now if such a value is inserted into Spanner?

Reply:
From the function definition, https://golang.org/pkg/math/big/#Rat.FloatString, it will always return a float string with the radix point. What do you mean by `such a value`? Do you mean an overflowed value, e.g., `0.1234567895`? Currently, we call `FloatString` so it will be rounded to `0.123456790` and then in spanner, it will be `0.123456790`.

Reply:
The if statement checks if the length of the slice is 2, which means it checks whether the string representation contains a point (`.`). So what I meant is:
- Is it at all possible that `strRep` will ever __not__ contain a `.`? If that is never possible, then this if statement is not necessary and can be removed.
- If it is possible, would it always be a problem if that value was inserted into Spanner? If for example the value is `100`, it should as far as I can tell be possible to insert the value into a `NUMERIC` column in Spanner. In that case the if check should probably check whether the number itself is too big to fit in a `NUMERIC` column, but it should not reject it as invalid.

Reply:
I check the length here to avoid the invalid index access (a runtime error) to the array later: 

```go
	whole := s[0]
	scale := s[1]
```

It's a just protection on our end. If the returned string is `100`, it means the `FloatString` does not work properly, which should not be ignored.

Reply:
Ok, that's reasonable I guess. My preference would be to remove the `if` statement in this case, as the contract of that function is to never return a string without a point. But if you feel it is better to keep it there, then I'm fine with that as well.

Reply:
You are right. We can actually remove the `if` because `FloatString` always returns a radix point as we specify to have a certain number of digits after the radix point and the missing numbers will be filled with zeros.